Spectral MACRO Score API Adapter

sluo10 opened the pull request:
This PR adds the Spectral MACRO Score Chainlink adapter. This adapter calls Spectral's MACRO Score API to retrieve a credit score and then computes a tick to determine the user's tick. For more information on Spectral's MACRO Score API and ticks, see docs.spectral.finance.

To test, run 

```bash
adapter=spectral-macro-score
yarn test $adapter/test/integration
```

Environment variables to properly run the adapter are in the README.

boxhock commented on packages/sources/spectral-macro-score/schemas/env.json:
```diff
@@ -0,0 +1,12 @@
+{
+  "$id": "https://external-adapters.chainlinklabs.com/schemas/example-source-adapter.json",
+  "title": "@chainlink/example-source-adapter env var schema",
+  "required": [],
+  "type": "object",
+  "properties": {},
```
Need to update id, title, required and properties. Here's an example: https://github.com/smartcontractkit/external-adapters-js/blob/develop/packages/sources/json-rpc/schemas/env.json

boxhock commented on packages/sources/spectral-macro-score/src/adapter.ts:
```diff
@@ -0,0 +1,44 @@
+  request: AdapterRequest,
+  config: SpectralAdapterConfig,
+): Promise<AdapterResponse> => {
+  console.log('execute')
+  const validator = new Validator(request, inputParams)
+  console.log('WUT')
+  if (validator.error) {
+    console.log(JSON.stringify(validator.error))
+    throw validator.error
+  }
```
:smile: 

```suggestion
  const validator = new Validator(request, inputParams)
  if (validator.error) {
    throw validator.error
  }
```

boxhock commented on packages/sources/spectral-macro-score/src/adapter.ts:
```diff
@@ -0,0 +1,44 @@
+  tickSetId: true,
+}
+
+export const execute = async (
+  request: AdapterRequest,
+  config: SpectralAdapterConfig,
+): Promise<AdapterResponse> => {
```
Here you should be able to make the constant of type `ExecuteWithConfig<SpectralAdapterConfig>`

boxhock replied:
Should also make room for a context object to be passed in.

sluo10 replied:
Can you explain why I need a context here? Documentation would help.

boxhock commented on packages/sources/spectral-macro-score/src/adapter.ts:
```diff
@@ -0,0 +1,44 @@
+  }
+}
+
+export const makeExecute: ExecuteFactory<SpectralAdapterConfig> = () => {
+  return async (request) => execute(request, makeConfig())
+}
```
Change with context support:

```suggestion
export const makeExecute: ExecuteFactory<SpectralAdapterConfig> = (config) => {
  return async (request, context) => execute(request, context, config || makeConfig())
}
```

boxhock commented on packages/sources/spectral-macro-score/src/config.ts:
```diff
@@ -0,0 +1,21 @@
+import { Requester, util } from '@chainlink/ea-bootstrap'
+import { Config } from '@chainlink/types'
+
+export const NAME = 'SPECTRAL_MACRO_SCORE' // This should be filled in with a name corresponding to the data provider using UPPERCASE and _underscores_.
```
```suggestion
export const NAME = 'SPECTRAL_MACRO_SCORE'
```

boxhock commented on packages/sources/spectral-macro-score/src/endpoint/macro-score-api.ts:
```diff
@@ -0,0 +1,69 @@
+import { getTickSet } from '../abi/NFC'
+import { SpectralAdapterConfig } from '../config'
+
+export const MacroScoreAPIName = 'spectral-proxy' // This should be filled in with a lowercase name corresponding to the API endpoint
```
```suggestion
export const MacroScoreAPIName = 'spectral-proxy'
```

boxhock commented on packages/sources/spectral-macro-score/src/endpoint/macro-score-api.ts:
```diff
@@ -0,0 +1,69 @@
+  const options = {
+    url: config.api,
+    method: 'POST',
+    data: `{"tokenInt":"${request.data.tokenIdInt}"}`,
```
I think we should be able to make this a proper object, and it will encode it for us - instead of making a stringified JSON object here.

boxhock commented on packages/sources/spectral-macro-score/src/endpoint/macro-score-api.ts:
```diff
@@ -0,0 +1,69 @@
+    data: `{"tokenInt":"${request.data.tokenIdInt}"}`,
+    headers: {
+      'Content-Type': 'application/json',
+      'x-api-key': config.apiKey ?? '',
```
Instead of setting API key here, consider setting the headers in the config object (`makeConfig()`), and import it here:
```javascript
{
  ...config.api,
  data: { [...] }
}
```

boxhock commented on packages/sources/spectral-macro-score/src/endpoint/macro-score-api.ts:
```diff
@@ -0,0 +1,69 @@
+      'Content-Type': 'application/json',
+      'x-api-key': config.apiKey ?? '',
+    },
+    timeout: 30000,
```
With the change above, you won't have to set timeout here. It will be read from the `API_TIMEOUT` env var (defaults to 30s)

boxhock commented on packages/sources/spectral-macro-score/src/endpoint/macro-score-api.ts:
```diff
@@ -0,0 +1,69 @@
+  const response = await Requester.request(options, customError)
+  response.data.result = Requester.validateResultNumber(response.data[0], ['score'])
+  const tick = computeTickWithScore(response.data[0].score, tickSet)
+  return Requester.success(request.data.jobRunID, { data: { result: tick } }, config.verbose)
```
See the suggestions I made in the README

boxhock reviewed (changes requested):
Thanks for the PR! :slightly_smiling_face:  Left some suggestions for improvements and some changes we need to be made before merging in.

sluo10 commented on .nvmrc:
```diff
@@ -0,0 +1 @@
+14.16.0
```
I'll remove this for this PR, but I would consider adding a `.nvmrc` file at some point. NVM is a common tool in the NodeJS ecosystem to help determine NodeJS and NPM versions for projects.

sluo10 commented on packages/sources/spectral-macro-score/README.md:
````diff
@@ -0,0 +1,58 @@
+```json
+{
+  "jobRunID": "278c97ffadb54a5bbb93cfec5f7b5503",
+  "data": [
+    {
+      "address": "0xa55E01a40557fAB9d87F993d8f5344f1b2408072",
+      "score_aave": "604.77",
+      "score_comp": "300.00",
+      "score": "452.38",
+      "updated_at": "2021-07-18T20:53:54.402553Z",
+      "is_updating_aave": false,
+      "is_updating_comp": false,
+      "result": 452.38
+    }
+  ],
````
I actually forgot to update this, it is no longer an array and instead is just returning an integer. Will update the README accordingly.

sluo10 commented:
@boxhock I addressed most of the comments, but am unsure how to proceed with some of them. For example, I don't fully understand why I need to pass in a context for the `makeExecute` function. If you think there is more work to be done let me know.

sluo10 commented:
@boxhock I updated the PR and I believe I've addressed everything! Thanks again for review 🙏🏼

thodges-gh commented:
Hey @sluo10 we just made some updates to our `develop` branch. Would you be able to rebase and push again, please?

sluo10 commented:
@thodges-gh thanks, updated!
